File: Dhruv/Microsoft/Day1/Combination Sum III.cpp
class Solution {
public:
    void CombinationSum(int ind, vector<int> &ds, vector<vector<int>> &ans, int target, int k, int n){
        if(ds.size()==k){
            if(target==0){
                ans.push_back(ds);
                return;
            }
        }
        for(int i=ind;i<=9;i++){
            if(ds.size()>=k && target!=0) break;
            if(i>target) break;
            ds.push_back(i);
            CombinationSum(i+1, ds, ans, target-i, k, n);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        vector<int> ds;
        if(n<k){
            return ans;
        }
        CombinationSum(1, ds, ans, n, k, n);
        return ans;
    }
};

File: Ravi/Microsoft/Day 1/CombinationSumIII.java
class Solution {
    
    List<List<Integer>> ans = new ArrayList<>();
    
    void combinationSum(int c, int i, List<Integer> curr, int n) {
        if(i == 0) {
            if(n == 0) ans.add(new ArrayList<>(curr));
            return;
        }
        
        for(int j = c + 1; j <= 9; j++) {
            curr.add(j);
            combinationSum(j, i - 1, curr, n - j);
            curr.remove(curr.size() - 1);
        }
        
    }
    
    public List<List<Integer>> combinationSum3(int k, int n) {
        
        List<Integer> curr = new ArrayList<>();
        combinationSum(0, k, curr, n);
        
        return ans;
    }
}
